Handles request rejection for expired subscriptions

sr-gi opened the pull request:
Expired subscriptions were not being checked when data was being sent by / requested from the tower. This PR handled that case and also moves the `UserInfo` deletion to the `Gatekeeper` instead of it being handled by either the `Watcher` or the `Responder`.

For this changes to work, the `Gatekeeper` now needs to have a notion of time, which is given by making him subscribe to the `ChainMonitor`. Also, data is being deleted with a `USERS_CACHE_SIZE_BLOCKS` delay so both the `Watcher` and the `Responder` can query any user related data they may need in order to do their own cleanups.

bigspider commented on teos/gatekeeper.py:
```diff
@@ -83,20 +89,46 @@ class Gatekeeper:
+
+    def manage_subscription_expiry(self):
+        """
+        Manages the subscription expiry of the registered users. When a subscription reaches it end, it moves it to
```
Typo, "its end"

sr-gi replied:
Changes the description here, so it does not apply anymore.

bigspider commented on teos/gatekeeper.py:
```diff
@@ -83,20 +89,46 @@ class Gatekeeper:
+    def manage_subscription_expiry(self):
+        """
+        Manages the subscription expiry of the registered users. When a subscription reaches it end, it moves it to
+        the expiry cache and deleted it after ``GATEKEEPER_USERS_CACHE`` blocks, giving time to the ``Watcher`` and
```
typo "deletes it", I suppose.

sr-gi replied:
Same as previous.

bigspider commented on teos/gatekeeper.py:
```diff
@@ -83,20 +89,46 @@ class Gatekeeper:
+                # data for the deletion of their data structures
+                if len(self.expired_users_cache) > USERS_CACHE_SIZE_BLOCKS:
+                    expired_users = self.expired_users_cache.pop(0)
+                    Cleaner.delete_expired_users(expired_users, self.registered_users, self.user_db)
```
Is this still safe in case of a block re-org? I'd think it is, as long USERS_CACHE_SIZE_BLOCKS is significantly bigger than what we consider the re-org threshold (e.g. 10 >> 6). Agreed?

sr-gi replied:
It is not actually, since the users in the cache are removed from `registered_users` without checking if the subscription is still expired after `USERS_CACHE_SIZE_BLOCKS`. Since the cache is not being reworked in a case of a reorg, it could be the case that it enters on a weird state. 

If we need to check `register_users` at `expiry + expiry_delta` and also `USERS_CACHE_SIZE_BLOCKS` blocks later, it may be way simpler to just drop items from the `register_users` at `expiry + expiry_delta + USERS_CACHE_SIZE_BLOCKS` and get rid of the cache.

That would mean that, effectively, we would be giving the user an additional `USERS_CACHE_SIZE_BLOCKS` windows where it may be able to renew the subscription and keep the data undeleted. I think that may be worth the simplicity of the approach tho.

sr-gi replied:
More on this. Since we are already giving `expiry_delta` additional blocks to the user, we could rework the way that works. Subscription could expire exactly at `expiry`, and the repurpose `expire_delta` as the time we keep information around before deleting. That means that the user won't be able to query no send any data from the tower after `expiry`, but their data will be kept around `expiry_delta` more blocks, and then it will be permanently deleted.

That gets rid both of the cache and the additional constant.

bigspider commented on teos/watcher.py:
```diff
@@ -294,10 +295,15 @@ def get_appointment(self, locator, user_signature):
+        if self.gatekeeper.has_subscription_expired(user_id):
+            raise SubscriptionExpired(
+                f"Your subscription expired at {self.gatekeeper.registered_users[user_id].subscription_expiry}"
+            )
```
Maybe better `"Your subscription expired at block {...}"`

sr-gi replied:
Fixed.

bigspider commented on teos/watcher.py:
```diff
@@ -347,6 +354,10 @@ def add_appointment(self, appointment, user_signature):
+        if self.gatekeeper.has_subscription_expired(user_id):
+            raise SubscriptionExpired(
+                f"Your subscription expired at {self.gatekeeper.registered_users[user_id].subscription_expiry}"
+            )
```
Same as above, I'd mention "block"

sr-gi replied:
Fixed.

bigspider commented on test/teos/unit/test_api.py:
```diff
@@ -189,8 +189,10 @@ def test_register_json_no_inner_dict(client):
 def test_add_appointment(internal_api, client, appointment, block_processor):
-    # Simulate the user registration (end time does not matter here)
-    internal_api.watcher.gatekeeper.registered_users[user_id] = UserInfo(available_slots=1, subscription_expiry=0)
+    # Simulate the user registration (end time does not matter here  as long as is in the future)
```
extra space, and typo "as long as it is"

sr-gi replied:
Fixed.

bigspider commented on test/teos/unit/test_gatekeeper.py:
```diff
@@ -196,6 +239,56 @@ def test_add_update_appointment(gatekeeper, generate_dummy_appointment):
+    user_id = get_random_value_hex(32)
+    gatekeeper.registered_users[user_id] = user_info
+
+    # Check tha the subscription is still live
```
typo "that"

sr-gi replied:
Fixed.

bigspider commented on test/teos/unit/test_gatekeeper.py:
```diff
@@ -196,6 +239,56 @@ def test_add_update_appointment(gatekeeper, generate_dummy_appointment):
+
+
+def test_get_expired_users(gatekeeper):
+    # Gets a list of users which subscription expires at a given height
```
typo "whose subscription"

sr-gi replied:
Fixed.

bigspider commented on test/teos/unit/test_watcher.py:
```diff
@@ -345,13 +347,32 @@ def test_add_appointment_no_slots(watcher, generate_dummy_appointment):
 
+# FIXME: 194 will do with dummy appointment
+def test_add_appointment_expired_subscription(watcher, generate_dummy_appointment):
+    # Appointments from register users with expired subscriptions fail as well
```
typo "registered users"

sr-gi replied:
Fixed.

bigspider reviewed (approved):
Only a bunch of typos.

sr-gi commented on teos/watcher.py:
```diff
@@ -338,7 +344,8 @@ def add_appointment(self, appointment, user_signature):
             :obj:`AppointmentLimitReached`: If the tower cannot hold more appointments (cap reached).
             :obj:`AuthenticationFailure`: If the user cannot be authenticated.
-            :obj:`NotEnoughSlots`: If the user does not have enough available slots, so the appointment is rejected.
+            :obj:`NotEnoughSlots`: If the user does not have enough available slots, so the appointment is rejected.\
```
Fixed.

sr-gi commented:
Reworked the subscription expiry to get rid of the cache as mentioned in https://github.com/talaia-labs/python-teos/pull/231#discussion_r492946714. I've added two separate commits that can be squashed once they're reviewed.

sr-gi commented:
Still missing tests for ``update_outdated_users_cache``, will add them tomorrow morning.

Needs rebasing to add all functionality in a single commit and all tests in another.

sr-gi commented:
Should be good now. Will need rebasing to add all functionality in a single commit and all tests in another once reviewed.

bigspider commented on teos/gatekeeper.py:
```diff
@@ -189,46 +228,117 @@ def add_update_appointment(self, user_id, uuid, ext_appointment):
 
-        return expired_appointments
+        if block_height not in self.outdated_users_cache:
+            with self.lock:
```
I would put all the code inside the lock, otherwise it could in principle happen that the code is stopped before getting the lock, meanwhile a reorg happens and a different block is processed, and the execution continues, while the state has changed. This would probably not be a problem here because the height after the re-org is likely the same, therefore the list of outdated users shouldn't change, but I think it's better to put the lock on the whole code and not have to think about it.

sr-gi replied:
Added

bigspider reviewed (changes requested):
Minor comments.
